### DISCUSSION TITLE
How to properly get orm mapped classes with eager loading from cte

### DISCUSSION THREAD

**Discussion by 20-buck-spin** [sqlalchemy/sqlalchemy#11781] [Usage Questions] [answered]
Hello, 

I would like to preface that I'm new to SqlAlchemy and am having trouble finding a suitable solution to the following problem. 

Let's suppose I have two related Classes Author and Book (book relates to author with a foreign key). 
I have a simple CTE expression  

`cte =  select(Author, Book).join(Book).cte()
`

I will be using this cte in a different part of the query. At the end of the query I would like to select the authors with their books loaded in the cte. I.e, select(Author).options(contains_eager(Author.books)).select_from(cte)

As of right now, I've been able to achieve this by joining Author and Books to the CTE again 

`
author =  (select(Author).
                      options(contains_eager(Author.books)).select_from(
                              cte.join(Author, cte.c.id == Author.id)
                                   .join(Book, Book.id == cte.c.book_id)
                      )
          )
`

This definitely works, but the resulting SQL first joins the Author and Book tables. and creates the necessary CTE. But at the end, it joins on the Author and Book tables again to get the necessary mapped classes. Is there someway to extract the necessary classes without joining a second time? Because all the necessary information is already in the cte. I was thinking maybe using a alias would help here. But I'm not sure if it's possible to create an alias for the tuple (Author, Book) which the cte returns. Any help here would be much appreciated! 

I am using a very simplified example, the real world query has many joins, creates another cte that uses a group by on the first cte, etc.

**Comment by CaselIT** [comment, marked as answer]
Hi,

You would need to use aliased:
```py
cte =  select(Author, Book).join(Book).cte() 

fa = orm.aliased(Author, cte)
fb = orm.aliased(Book, cte)

s = sa.select(fa).options(orm.contains_eager(fc.books.of_type(fb))).select_from(cte)
```

**Comment by 20-buck-spin** [comment]
Thank you very much!!! It works like a charm. I wasn't aware of the of_type construct.
